gadget/mountedfilesystem.go

```diff
@@ -0,0 +1,186 @@
+	}
+	fw := &MountedFilesystemWriter{
+		contentDir: contentDir,
+		ps:         ps,
+	}
+	return fw, nil
+}
+
+func prefixPreserve(dstDir string, preserve []string) []string {
+	preserveInDst := make([]string, len(preserve))
+	for i, p := range preserve {
+		preserveInDst[i] = filepath.Join(dstDir, p)
+	}
+	sort.Strings(preserveInDst)
+
+	return preserveInDst
+}
+
+// Write writes structure data into provided directory. All existing files are
+// overwritten, unless their paths, relative to target directory, are listed in
+// the preserve list.
+func (m *MountedFilesystemWriter) Write(whereDir string, preserve []string) error {
+	if whereDir == "" {
+		return fmt.Errorf("internal error: destination directory cannot be unset")
+	}
+	preserveInDst := prefixPreserve(whereDir, preserve)
+	for _, c := range m.ps.Content {
+		if err := m.writeVolumeContent(whereDir, &c, preserveInDst); err != nil {
+			return fmt.Errorf("cannot write filesystem content of %s: %v", c, err)
+		}
+	}
+	return nil
+}
+
+// writeDirectory copies the source directory, or its contents under target
+// location dst. Follows rsync like semantics, that is:
+//   /foo/ -> /bar - deploys contents of foo under /bar
+//   /foo  -> /bar - deploys foo and its subtree under /bar
+func writeDirectory(src, dst string, preserveInDst []string) error {
+	hasDirSourceSlash := strings.HasSuffix(src, "/")
+
+	if !osutil.IsDirectory(src) {
+		if hasDirSourceSlash {
+			// make the error sufficiently descriptive
+			return fmt.Errorf("cannot specify trailing / for a source which is not a directory")
+		}
+		return fmt.Errorf("source is not a directory")
+	}
+
+	if !hasDirSourceSlash {
+		// /foo -> /bar (deploy foo and subtree)
+		dst = filepath.Join(dst, filepath.Base(src))
+	}
+
+	fis, err := ioutil.ReadDir(src)
+	if err != nil {
+		return fmt.Errorf("cannot list directory entries: %v", err)
+	}
+
+	for _, fi := range fis {
+		pSrc := filepath.Join(src, fi.Name())
+		pDst := filepath.Join(dst, fi.Name())
+
+		write := writeFile
+		if fi.IsDir() {
+			if err := os.MkdirAll(pDst, 0755); err != nil {
+				return fmt.Errorf("cannot create directory prefix: %v", err)
+			}
+
+			write = writeDirectory
+			pSrc += "/"
+		}
+		if err := write(pSrc, pDst, preserveInDst); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// writeFile copies the source file at given location or under given directory.
+// Follows rsync like semantics, that is:
+//   /foo -> /bar/ - deploys foo as /bar/foo
+//   /foo  -> /bar - deploys foo as /bar
+// The destination location is overwritten.
+func writeFile(src, dst string, preserveInDst []string) error {
+	if strings.HasSuffix(dst, "/") {
+		// deploy to directory
+		dst = filepath.Join(dst, filepath.Base(src))
+	}
+
+	if osutil.FileExists(dst) && strutil.SortedListContains(preserveInDst, dst) {
+		// entry shall be preserved
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Dir(dst), 0755); err != nil {
```

Comment:
Is there any plan to preserve directory permissions during the copy operation?

Reply:
As discussed on IRC, I'll document that and leave a TODO in the code